File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let set = new Set();
    let i = 0,j=0, ans = 0;
    if(s.length == 0) return 0;
    if(s.length == 1) return 1;
    
    while(j<s.length)
        {
            while(set.has(s[j]))
                {
                    set.delete(s[i]);
                    i++;
                }
            set.add(s[j]);
            ans = Math.max(ans,j-i+1);
            j++;
        }
    return ans;
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int i=0,j=0,len = s.size(), ans = INT_MIN;
        if(len == 0) return 0;
        unordered_map<char,int>m;
        while(j<len)
        {
            //normal calc
            m[s[j]]++;
            if(m.size() == j-i+1)
            {
                //calc of ans
                ans = max(ans,j-i+1);
            }
            else if(m.size()<j-i+1)
            {
                while(m.size()<j-i+1)
                {
                    m[s[i]]--;
                    if(m[s[i]]==0)
                    {
                        m.erase(s[i]);
                    }
                    i++;
                }
            }
            j++;
        }
        return ans;
    }
};